### ISSUE TITLE
Obtain all User Saved Posts or Upvoted posts via Reddit Data Request OR through Pushshift if possible!

### ISSUE THREAD

**Issue by Kristijan1001** [shadowmoose/RedditDownloader#134]
### Is your feature request related to a problem? 
Reddit API Limit.

### Describe the solution you'd like:
Use the "saved_posts.csv" from reddit data request to obtain old saved/upvoted posts that are no longer obtainable through the normal Reddit API, or use Pushshift if there is a way to get them through that.

Reddit Request Data Feature:

https://www.reddit.com/settings/data-request

### Additional context:
There is a topic on this on pushshift subreddit and looks like someone already made a script to obtain all the saved posts, but not the upvoted one. Also it is not user friendly since it is just a script. Link to the reddit thread:

https://www.reddit.com/r/pushshift/comments/gbsxyi/would_it_be_possible_to_get_all_posts_saved_on_an/
Code from the Kanra951 user in the reddit thread in case it gets removed:

```

import requests
import json
import concurrent.futures
import os
import csv
import re

def getSaved(link, saved):
    url = f'{link}.json'
    resp = requests.get(url, headers = {'User-agent': 'lol'})
    if resp.ok:
        data = resp.json()[0]["data"]["children"][0]["data"]
        try:
            if data["domain"] == "gfycat.com":
                filename = f'{data["created_utc"]}_{data["title"]}'
                filename = re.sub('[^A-Za-z0-9_ ]+', '', filename)
                if len(filename) > 50:
                    filename = filename[:50]
                c = {
                    "domain": "gfycat",
                    "title": data["title"],
                    "permalink": f'https://reddit.com{data["permalink"]}',
                    "score": data["score"],
                    "subreddit": data["subreddit_name_prefixed"],
                    "url": data["url"],
                    "gfycaturl": data["url"].split('/')[-1].split('-')[0],
                    "thumbnail": data["thumbnail"],
                    "timestamp": data["created_utc"],
                    "filename": f'{filename}.mp4'
                }
                saved.append(c)
            elif data["domain"] == "i.imgur.com":
                if data["url"].split('.')[-1] == "gif" or data["url"].split('.')[-1] == "gifv":
                    filename = f'{data["created_utc"]}_{data["title"]}'
                    filename = re.sub('[^A-Za-z0-9_ ]+', '', filename)
                    if len(filename) > 50:
                        filename = filename[:50]
                    c = {
                        "domain": "imgur",
                        "title": data["title"],
                        "permalink": f'https://reddit.com{data["permalink"]}',
                        "score": data["score"],
                        "subreddit": data["subreddit_name_prefixed"],
                        "url": data["url"],
                        "imgururl": data["url"].split('/')[-1].split('.')[0],
                        "thumbnail": data["thumbnail"],
                        "timestamp": data["created_utc"],
                        "filename": f'{filename}.mp4'
                    }
                    saved.append(c)
        except Exception as e:
            print(f'Get saved exception: {e}')

def downloadGfy(i, directory):
    if i["domain"] == "gfycat":
        gfycatAPI = f'https://api.gfycat.com/v1/gfycats/{i["gfycaturl"]}'
        link = requests.get(gfycatAPI).json()["gfyItem"]["mp4Url"]
        r = requests.get(link, stream = True)
        with open(f'{directory}/{i["filename"]}', 'wb') as f: 
            for chunk in r.iter_content(chunk_size = 1024*1024): 
                if chunk: 
                    f.write(chunk)
    elif i["domain"] == "imgur":
        link = f'http://i.imgur.com/{i["imgururl"]}.mp4'
        r = requests.get(link, stream = True)
        with open(f'{directory}/{i["filename"]}', 'wb') as f: 
            for chunk in r.iter_content(chunk_size = 1024*1024): 
                if chunk: 
                    f.write(chunk)

links = []
with open('saved_posts.csv', 'r') as f:
    reader = csv.reader(f)
    for row in reader:
        if row[1] != "permalink":
            links.append(row[1])
print("Links no: ", len(links))

saved = []
with concurrent.futures.ThreadPoolExecutor(max_workers=100) as executor:
    futures = {executor.submit(getSaved, i, saved): i for i in links}
    for future in concurrent.futures.as_completed(futures):
        try:
            data = future.result()
        except Exception as threadExp:
            print(threadExp)

# Sort by date the post was created
saved = sorted(saved, key=lambda k: k['timestamp'])
saved.reverse()

# Generate html
with open("index.html", "w", encoding="utf-8") as f:
    html = f'''<!DOCTYPE html>
    <head></head>
    <body>
    <style>.comment {{margin: 30px 50px 30px;}}
    img{{width:100%;max-width:300px;}}
    h1 {{margin: 10px 25px 10px;}} 
    .divider {{border-top: 1px solid #bbb;}}
    p {{margin-block-start: 0px;
    margin-block-end: 0px;}}
    .top {{font-size:12px}}
    body {{font-family: 'Helvetica', 'Arial', sans-serif;}}</style>
    <div>
    <div class="comment">'''
    for i in range(0, len(saved)):
        html = html + f'''
        <p class="top"><b>{saved[i]["subreddit"]}</b> {saved[i]["score"]} points <a href="{saved[i]["permalink"]}">permalink</a></p>
        <p>{saved[i]["title"]} <a href="{saved[i]["url"]}">{saved[i]["url"]}</a></p>
        <div><a href="{saved[i]["url"]}"><img src="{saved[i]["thumbnail"]}"></a></div>
        <hr class="divider">
        '''
    html = html + "</div></div></body>"
    f.write(html)
print("Complete. Results saved in index.html")
os.system("start " + "index.html")

# Download gfycats and imgur vids
directory = './gfy'
if not os.path.exists(directory):
    os.makedirs(directory)
with concurrent.futures.ThreadPoolExecutor(max_workers=100) as executor:
    futures = {executor.submit(downloadGfy, i, directory): i for i in saved}
    for future in concurrent.futures.as_completed(futures):
        try:
            data = future.result()
        except Exception as threadExp:
            print(threadExp)

print(len(saved))
reddit-download-saved-videos.py
Displaying reddit-download-saved-videos.py.
```

**Comment by shadowmoose** [comment]
This is possible, but I'll need a sample file first. I've requested one, so whenever it arrives I'll look into this.

**Comment by Kristijan1001** [comment]
@shadowmoose I just received mine it took less than 12 hours. It contains separate files, which has all of the links u have upvoted and saved with no limit. It is a CSV file. I will attach an example of how it looks. It is super simply laid out 1 link per line. I can't upload the csv file directly so here is a .zip file.


[post_votes & saved_posts.zip](https://github.com/shadowmoose/RedditDownloader/files/4625066/post_votes.saved_posts.zip)
![EXCEL_ptI3iB5Qet](https://user-images.githubusercontent.com/24441639/81872652-703fce80-957a-11ea-8bc2-66de40028381.png)
![EXCEL_FaEMXUYgQ6](https://user-images.githubusercontent.com/24441639/81872681-80f04480-957a-11ea-81e1-e9241fc1a33a.png)

**Comment by shadowmoose** [comment]
Thanks for the suggestion, and the data. I've implemented this as of 85ff9ca650e66bc6d7f4fc2ee590c8826540bc1b in the master branch. RMD will accept a CSV file path argument, which it will then scan for Posts. 

It will be released sometime soon with RMD 3.1.0.

**Comment by Kristijan1001** [comment]
> Thanks for the suggestion, and the data. I've implemented this as of [85ff9ca](https://github.com/shadowmoose/RedditDownloader/commit/85ff9ca650e66bc6d7f4fc2ee590c8826540bc1b) in the master branch. RMD will accept a CSV file path argument, which it will then scan for Posts.
> 
> It will be released sometime soon with RMD 3.1.0.

Is it possible to get an up to date early build of the program ?

**Comment by shadowmoose** [comment]
I have release RMD 3.1.0 as of now, so this feature should be available everywhere.
